### PR TITLE
Split modules between NodeJS and browser

### PR DESCRIPTION
**Pull request by emizzle**
First version of a swarm js library for interacting with a swarm gateway

The library is transpiled by babel to work with Node and the browser. The `uploadDirectory` function is only available in Node (due to the `fs` requirement).

Still left to do:
- Integrate with ENS so uploaded content can be accessed via an ENS name instead of a hash
- Modify `uploadDirectory` function so that the `defaultPath` can be specified (currently it doesn't work)
- unit tests

Related to https://github.com/embark-framework/embark/pull/802.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,34 +1,23 @@
...
     let readables = {}; // files, directories, symlinks, etc
     let errors = [];
 
     const excludeDirFilter = through2.obj(function (item, enc, next) {
-      if (!item.stats.isDirectory()) this.push(item);
-      next();
-    });
+      if (!item.stats.isDirectory()) this.push(item)
+      next()
+    })
```
Comment (jrainville) [inline src/index.js]: Why remove the `;`?

File: src/shared.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+class _SwarmJS {
+
+  constructor(opts) {
+    this.opts = opts || {};
+    if (this.opts.gateway) {
+      this.gateway = opts.gateway
+    } else if (this.opts.mode === 'http') {
+      this.gateway = 'http://swarm-gateways.net'
+    } else {
+      this.gateway = 'https://swarm-gateways.net'
```
Comment (jrainville) [inline src/shared.js]: You have swarm-gateway twice. Maybe you can remove the `else if` then?

**Review by jrainville — APPROVED** [review]

File: swarmjs.min.js [hunk 3]
```diff
@@ -0,0 +1 @@
...pe,e),n&&u(t,n)}(t,[{key:"_isValidHash",value:function(t){return/^[0-9a-f]{64}$/.test(t)}},{key:"_hashResponse",value:function(t,e,n,r){t?r(t):200!==e.statusCode?r(n):this._isValidHash(n)?r(null,n):r("Invalid hash")}},{key:"download",value:function(t,e){i("".concat(this.gateway,"/").concat(t),function(t,n,r){t?e(t):200!==n.statusCode?e(r):e(null,r)})}},{key:"downloadRaw",value:function(t,e){this.download("bzz-raw:/".concat(t),e)}},{key:"upload",value:function(t,e,n){var r=this;i.post({url:"".concat(this.gateway,"/").concat(t),body:e},function(t,e,o){return r._hashResponse(t,e,o,n)})}},{key:"uploadRaw",value:function(t,e){this.upload("bzz-raw:/",t,e)}},{key:"uploadForm",value:function(t){var e=this,n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:"index.html",r=arguments.length>2?arguments[2]:void 0,o={url:"".concat(this.gateway,"/bzz:/"),formData:t,qs:{defaultpath:n}};i.post(o,function(t,n,o){return e._hashResponse(t,n,o,r)})}},{key:"isAvailable",value:function(t){try{this.uploadRaw("test",function(e,n){if(e)return t(e);t(null,"6de1faa7d29b1931b4ba3d44befcf7a5e43e947cd0bf2db154172bac5ecac3a6"===n)})}catch(e){t(e)}}}]),t}();n.d(e,"default",function(){return a})}]).default});
```
Comment (alaibe) [inline swarmjs.min.js]: Is this file supposed to be version controlled?

File: webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+const path = require('path');
+
+const standalone = {
+    entry: path.join(__dirname, 'dist/browser', 'browser.js'),
```
Comment (alaibe) [inline webpack.config.js]: maybe keep 2 spaces indentation everywhere?

File: src/index.js (hunk 1 above)
Comment (emizzle) [reply src/index.js]: Good catch, this was a rebase problem. Will fix.

File: src/shared.js (hunk 2 above)
Comment (emizzle) [reply src/shared.js]: Haha I knew it would trigger you 😜 I'll extract the `swarm-gateways.net`.

File: swarmjs.min.js (hunk 3 above)
Comment (emizzle) [reply swarmjs.min.js]: Good point, I must've added it to the .gitignore after it was already added or possibly got readded in a rebase. Will remove from the index.

File: webpack.config.js (hunk 4 above)
Comment (emizzle) [reply webpack.config.js]: 100%, good catch

**Pull request merged**